core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/StrictEvaluationStrategy.java

```diff
@@ -730,52 +730,48 @@ protected Value getVarValue(Var var, BindingSet bindings) {
 
 	public CloseableIteration<BindingSet, QueryEvaluationException> evaluate(BindingSetAssignment bsa,
 			BindingSet bindings) throws QueryEvaluationException {
-		final Iterator<BindingSet> iter = bsa.getBindingSets().iterator();
-		if (bindings.size() == 0) { // empty binding set
-			return new CloseableIteratorIteration<>(iter);
+		final Iterator<BindingSet> assignments = bsa.getBindingSets().iterator();
+		if (bindings.size() == 0) {
+			// we can just return the assignments directly without checking existing bindings
+			return new CloseableIteratorIteration<>(assignments);
 		}
 
+		// we need to verify that new binding assignments do not overwrite existing bindings
 		CloseableIteration<BindingSet, QueryEvaluationException> result;
 
-		final QueryBindingSet b = new QueryBindingSet(bindings);
-
 		result = new LookAheadIteration<BindingSet, QueryEvaluationException>() {
 
 			@Override
 			protected BindingSet getNextElement() throws QueryEvaluationException {
-				QueryBindingSet result = null;
-				while (result == null && iter.hasNext()) {
-					final BindingSet assignedBindings = iter.next();
+				QueryBindingSet nextResult = null;
+				while (nextResult == null && assignments.hasNext()) {
+					final BindingSet assignedBindings = assignments.next();
+
 					for (String name : assignedBindings.getBindingNames()) {
+						if (nextResult == null) {
+							nextResult = new QueryBindingSet(bindings);
+						}
+
```

Comment:
This is the crux of the fix: we initialize `nextResult` as soon as we have a binding name (previously it was only initialized if that binding name also had a value in the BSA)